File: python/2015/day01b/day01b.py
#!/usr/bin/env python3

import sys

def usage():
    progname = sys.argv[0]
    print(f'usage: {progname} <input file>')
    sys.exit(1)

def process(filename):
    counter = 0
    pos = 0
    with open(filename, 'r') as infile:
        for line in infile:
            for ch in line:
                pos += 1
                if ch == '(':
                    counter += 1
                elif ch == ')':
                    counter -= 1
                if counter < 0:
                    return pos
    return 0

def main():
    if len(sys.argv) < 2:
        usage()
    filename = sys.argv[1]
    result = process(filename)
    print(f'result = {result}')

if __name__ == '__main__':
    main()

File: perl/2015/day01b/day01b.pl
#!/usr/bin/env perl

use strict;
use warnings;

sub usage {
    my $progname = $0;
    print "usage: $progname <input file>\n";
    exit 1;
}

sub process {
    my ($filename) = @_;
    my $counter = 0;
    my $pos = 0;

    open ( my $infile, '<', $filename ) or die "cannot open file: $filename: $!";
    while ( my $line = <$infile> ) {
        foreach my $ch ( split //, $line ) {
            $pos++;
            if ( $ch eq '(' ) {
                $counter++;
            } elsif ( $ch eq ')' ) {
                $counter--;
            }
            if ( $counter < 0 ) {
                return $pos;
            }
        }
    }
    close($infile);

    return 0;
}

sub main {
    if ( @ARGV < 1 ) {
        usage();
    }
    my $filename = $ARGV[0];
    my $result = process($filename);
    print "result = $result\n";
}

main();
